File: images6/publish.py
import logging
import bottle
from datetime import datetime

from .system import current_system
from .types import PropertySet, Property, EnumProperty
from .metadata import register_metadata_schema
from .web import App as MainApp


# WEB
#####


class App:
    BASE = '/publish'

    @classmethod
    def create(self):
        app = bottle.Bottle()

        app.route(
            path='/',
            callback=lambda: bottle.static_file(
                'publish.html', root=MainApp.HTML
            ),
        )

        return app


# DATAMODEL
###########

class ElementType(EnumProperty):
    image = 'image'
    text = 'text'
    heading = 'heading'


class Element(PropertySet):
    type = Property(enum=ElementType, default=ElementType.image)
    src = Property()
    content = Property()
    entry_id = Property()


class Page(PropertySet):
    number = Property(int)
    show = Property(bool, default=True)
    width = Property(int)
    height = Property(int)
    elements = Property(type=Element, is_list=True)


class Publication(PropertySet):
    author = Property()
    publish_ts = Property()
    title = Property()
    comment = Property()
    pages = Property(type=Page, is_list=True)


register_metadata_schema(Publication)

File: web/js/publish.js
$(function() {
    var Publication = function(args) {
        var args = args || {};
        var me = {
            current_page: 0,
        };

        me.entry = {
            metadata: {
                '*Schema': 'Publication',
                title: args.title || 'untitled',
                author: args.author || null,
                comment: args.comment || null,
                publish_ts: args.publish_ts || null,
                pages: [],
            },
        };

        me.create_page = function(args) {
            var set_as_current = args.set_as_current || false;
            var page = Page(args);
            var new_length = me.entry.metadata.pages.push(page);
            if (set_as_current) {
                me.current_page = new_length - 1;
            }
            return page;
        };

        return me;
    };
    
    var Page = function(args) {
        var args = args || {};
        var me = {
            '*Schema': 'Page',
        };

        me.number = args.number === undefined ? 0 : args.number;
        me.show = args.show === undefined ? true : args.show;
        me.width = args.width === undefined ? 80 : args.width;
        me.height = args.height === undefined ? 50 : args.height;
        me.elements = args.elements === undefined ? [] : args.elements;

        me.create_element = function(args) {
            var element = Element(args);
            me.elements.push(element);
            return element;
        };

        return me;
    };

    var Element = function(args) {
        var args = args || {};
        var me = {
            '*Schema': 'Element',
        };

        me.type = args.type || 'image';
        me.src = args.src || null;
        me.content = args.content || null;
        me.entry_id = args.entry_id || null;

        return me;
    };

    $.Images = $.Images || {};
    $.Images.Publication = Publication;
    $.Images.Page = Page;
    $.Images.Element = Element;
});
